### ISSUE TITLE
Index defined Resources, Components, Systems and Bundles in the Book or in the documentation

### ISSUE THREAD

**Issue by ldesgoui** [amethyst/amethyst#986]
Hello,

Currently, if I were to search for all Components defined by Amethyst, I would have to look through all the crates' documentations, that's not exactly the best experience.
What I'd really like is a page where all the Resources, Components, etc, are indexed and categorized to help developers find what they need.
This was discussed in Discord: https://discordapp.com/channels/425678876929163284/425678876929163287/496401358157512704

Thanks

**Comment by ldesgoui** [comment]
@Moxinilian wants a tag

**Comment by VermillionAzure** [comment]
Given Amethyst's docs are currently created as a static website with `rustdoc`, perhaps we may want to consider an extension to `rustdoc` or leverage some of it. Some investigation into this may be needed. Let me see what I can do.

**Comment by VermillionAzure** [comment]
Looks like [`rustdoc` itself is intertwined with the compiler itself](https://rust-lang-nursery.github.io/rustc-guide/rustdoc.html)?

**Comment by Moxinilian** [comment]
I think something similar to the Rust nomicon might fulfill that task. An  encyclopedia-like book for people that now know how the basics of the engine works and want a list of small articles on every single features of the engine.

**Comment by VermillionAzure** [comment]
@ldesgoui 

1. Can you explain more about what you want as a user experience?

    What I'd really like is a page where all the Resources, Components, etc, are indexed and categorized to help developers find what they need.

2. You said "index" -- do you want something stronger than [this type of search in the current static docs](https://www.amethyst.rs/doc/latest/doc/amethyst/?search=Resource)?

**Comment by ldesgoui** [comment]
The search doesn't really do it justice because of all the results that come up, the UX doesn't need to be overcomplicated, literally this would work:

| Name |  Description |
| --- | --- |
| [Time (Resource)](#) | Linear time since program was launched |
| [TimeSystem](#) |  Updates the Time resource |
| [TimeBundle](#) | Sets up Time and TimeSystem |

(Links would point to docs page obviously)

As mentioned, it would take 5-10 minutes of checking up every time someone commits/PRs but I think the benefits are quite important

**Comment by Telzhaak** [comment]
Maybe this helps a little:
gathered by @taqueso (from Discord), dated August 19th, 2018

systems:
TransformSystem                 amethyst_core/src/transform/systems.rs
Processor                       amethyst_assets/src/storage.rs
PrefabLoaderSystem              amethyst_assets/src/prefab/system.rs
HotReloadSystem                 amethyst_assets/src/reload.rs
RenderingSystem                 amethyst_assets/examples/hl.rs
UiKeyboardSystem                amethyst_ui/src/text.rs
ResizeSystem                    amethyst_ui/src/resize.rs
UiMouseSystem                   amethyst_ui/src/event.rs
UiTransformSystem               amethyst_ui/src/layout.rs
UiButtonSystem                  amethyst_ui/src/button/system.rs
SpriteVisibilitySortingSystem   amethyst_renderer/src/sprite_visibility.rs
VisibilitySortingSystem         amethyst_renderer/src/visibility.rs
InputSystem                     amethyst_input/src/system.rs
FPSCounterSystem                amethyst_utils/src/fps_counter.rs
NopSystem                       src/game_data.rs
DjSystem                        amethyst_audio/src/systems/dj.rs
AudioSystem                     amethyst_audio/src/systems/audio.rs
SamplerInterpolationSystem      amethyst_animation/src/systems/sampling.rs
AnimationControlSystem          amethyst_animation/src/systems/control.rs
VertexSkinningSystem            amethyst_animation/src/skinning/systems.rs
FlyMovementSystem               amethyst_controls/src/systems.rs
ArcBallMovementSystem           amethyst_controls/src/systems.rs
FreeRotationSystem              amethyst_controls/src/systems.rs
MouseFocusUpdateSystem          amethyst_controls/src/systems.rs
CursorHideSystem                amethyst_controls/src/systems.rs

bundles:
TransformBundle       amethyst_core/src/transform/bundle.rs
HotReloadBundle       amethyst_assets/src/reload.rs
UiBundle              amethyst_ui/src/bundle.rs
InputBundle           amethyst_input/src/bundle.rs
FPSCounterBundle      amethyst_utils/src/fps_counter.rs
AudioBundle           amethyst_audio/src/bundle.rs
VertexSkinningBundle  amethyst_animation/src/bundle.rs
SamplingBundle        amethyst_animation/src/bundle.rs
AnimationBundle       amethyst_animation/src/bundle.rs
FlyControlBundle      amethyst_controls/src/bundles.rs
ArcBallControlBundle  amethyst_controls/src/bundles.rs

components:
Named                 amethyst_core/src/named.rs
GlobalTransform       amethyst_core/src/transform/components/transform.rs
Transform             amethyst_core/src/transform/components/local_transform.rs
Parent                amethyst_core/src/transform/components/parent.rs
Handle                amethyst_assets/src/storage.rs
PrefabTag             amethyst_assets/src/prefab/mod.rs
OnUiActionImage       amethyst_ui/src/action_components.rs
OnUiActionSound       amethyst_ui/src/action_components.rs
UiImage               amethyst_ui/src/image.rs
UiText                amethyst_ui/src/text.rs
TextEditing           amethyst_ui/src/text.rs
UiTransform           amethyst_ui/src/transform.rs
UiResize              amethyst_ui/src/resize.rs
MouseReactive         amethyst_ui/src/event.rs
UiButton              amethyst_ui/src/button/mod.rs
Light                 amethyst_renderer/src/light.rs
Camera                amethyst_renderer/src/cam.rs
Material              amethyst_renderer/src/mtl.rs
SpriteRender          amethyst_renderer/src/sprite.rs
JointTransforms       amethyst_renderer/src/skinning.rs
Transparent           amethyst_renderer/src/transparent.rs
Tag                   amethyst_utils/src/tag.rs:impl<T> Component
AudioListener         amethyst_audio/src/components/audio_listener.rs
AudioEmitter          amethyst_audio/src/components/audio_emitter.rs
RestState             amethyst_animation/src/resources.rs
AnimationHierarchy      "
SamplerControlSet       "
AnimationControl        "
AnimationControlSet     "
AnimationSet            "
Joint                 amethyst_animation/src/skinning/resources.rs
Skin                    "
FlyControlTag         amethyst_controls/src/components.rs
ArcBallControlTag       "
GltfNodeExtent        amethyst_gltf/src/lib.rs

**Comment by jchitel** [comment]
It would definitely be super valuable to have sections of the book dedicated as a "catalog" of systems, components, and bundles provided out of the box by the engine, complete with short descriptions of what they are used for and links to the docs. At least just for the common ones.

I'm brand new to ECS in general. I just read the Specs docs and then discovered Amethyst. Since it's so new to me, I wanted to see some examples of what real world components and systems look like, but couldn't find this in the book, and it is hard to track down in the docs. I have to imagine other newcomers would also be interested in seeing an explicit list of the built-in Amethyst ones.

As a general comment on the book, I think it would be prudent to highlight specific "commonly used" APIs that developers will almost always need to use in a fully-featured game. You can get a lot from the generated docs, but it can be hard to sift through and find those crucial APIs.

I find that a great way to learn a tool with insufficient documentation is to learn what you need to add that documentation yourself! So I'd be happy to take this on if I could be pointed in the right direction of what qualifies as "crucial" or "commonly used" (as opposed to "rarely used" APIs that developers will likely only need for edge cases).

**Comment by torkleyy** [comment]
> As a general comment on the book, I think it would be prudent to highlight specific "commonly used" APIs that developers will almost always need to use in a fully-featured game. You can get a lot from the generated docs, but it can be hard to sift through and find those crucial APIs.

I think that's a good idea!

> So I'd be happy to take this on if I could be pointed in the right direction of what qualifies as "crucial" or "commonly used" (as opposed to "rarely used" APIs that developers will likely only need for edge cases).

Thank you for offering your help! I think this is a bit hard to point to. There are not that many games, so we don't have very much data (although we are slowly getting to the point). I think we have to collect that on some thread and gradually build up the catalog.

I have one concern. Right now our book is a bit a mix ot tutorial-style and reference style. Then there's also the API docs. Before building a catalog, we should carefully think about the interaction.

Duplication is something that should be avoided in my opinion. We need to focus on having one place for documentation for each target audience instead of many covering things partially or with poor quality. There are definitely some intersections; those can be covered by linking (maybe to the API docs?).

One thing I can imagine for the catalog is to only provide a very short explanation and to only link to the important concepts. In that case, I suggest we call the catalog "cheat sheet". I'm not opposed to other ideas though.

So to summarize, I'd like to assist you with this, but I'm having a hard time to link you to the relevant places right now. We also first need to commit to an overall strategy to ensure some integrity. After that, we can open a collection thread where you and / or others can incorporate suggested "commonly used scenarios" into such a catalog.

**Comment by jojolepro** [comment]
Also having a list of ALL systems, resources and components could be beneficial for more advanced users.  It would just be a list of links pointing to the docs.

The important thing to remember for this is that it should be automated.

**Comment by qqwa** [comment]
I also thought about something like that and started a PR currently just a [commit](https://github.com/qqwa/amethyst/commit/bd8174d706fe7b8e1ffed6164f1e3285466b36ac)

But then I found the relevant parts in the doc myself, so I'm not so sure about it anymore, because of the duplication part @torkleyy mentioned.

For example: [specs::Systems](https://www.amethyst.rs/doc/master/doc/specs/trait.System.html#implementors) here we can find all Systems, only draw back is that there is no inline description. Another draw back is that other references like [shred::System](https://www.amethyst.rs/doc/master/doc/shred/trait.System.html) don't show any implementors. So that needs to be fixed somehow, maybe by just using `#[doc(hidden)]`

If we use the [search function](https://www.amethyst.rs/doc/master/doc/specs/index.html?search=System) we get Systems with a short description. But a lot of "false positives". Which can be minimized by building the docs with `cargo doc --no-deps -p amethyst -p amethyst_assets ...` This only is useful for Systems and Bundles, as Resources don't have Resource in its name i think. But it should still be useful because we have in general less clutter when searching the docs.

Another thing that is really hard to find are Type Definitions like [Entities](https://www.amethyst.rs/doc/master/doc/specs/type.Entities.html). You can't discover Type Definitions that wrap a particular struct/trait. So no idea how we would solve that inside the api docs. Especially because everything is splitted between so much crates.

A Reference/Catalog chapter in the book has the advantage that it is really easy to discover stuff, but is extra work either to hold it up to date or to write the tool to do it automatically. Another advantage would be that we could also link systems etc. of extern projects.

Doing it with the rust api docs is in my opinion the "correct" way, if we can figure out how to make systems, resources etc. easier to discover within the docs. The biggest problem is probably that we use multiple crates and how to handle the documentation between them. Easiest way would be to just add a list to book..

**Comment by jchitel** [comment]
So, API docs in any language serve a specific purpose: providing a technical listing of the full public interface of the library organized by module, using a standard format that all users of the language know how to navigate. We could certainly look into extending that, but in that case we are no longer serving the "standard format" requirement of the API docs.

In my opinion, everything that cannot be satisfied by API docs belongs in "custom documentation", which in this case is the book.

The most successful organization of documentation that I've seen is as follows:
1. Begin with an introduction to the tool in question and outline the structure of the docs. [This exists in the right place](https://www.amethyst.rs/book/master/intro.html).
2. Follow with a "quick start" section, including instructions on installing the tool and getting a "hello world" example running. [This exists in the right place](https://www.amethyst.rs/book/master/getting-started.html).
2. Next, a tutorial that walks through a basic implementation using the tool. This should touch on all of the fundamentals, how to structure a project, and provide at least an introduction of how the tool works internally, but most importantly it should focus on *what to do* to get a working practical example running, linking to other sections of the book for more details. [This is currently being served by the Pong tutorial, but could use a little more (beyond actually finishing it)](https://www.amethyst.rs/book/master/pong-tutorial.html). One important thing required here is establishing required knowledge right at the beginning of the tutorial. For the pong tutorial, I'd say that the only "required" knowledge is a working understanding of Rust, and perhaps a little knowledge of what goes into making a game. Having established that, the tutorial should be written as if that is the only thing that the reader knows. I feel that the tutorial is mostly there, but could use a little more description in certain places. The other problem is that this should immediately follow the "getting started" section, but the book instead has a [concepts](https://www.amethyst.rs/book/master/concepts/intro.html) section in between, and I don't think that is a particularly good place for that. Additionally, the [Appendix A](https://www.amethyst.rs/book/master/appendices/a_config_files.html) section definitely belongs as a subsection of the tutorial, and isn't really an "appendix".
4. Following the tutorial should be several chapters (or a single large section) that serve as a more technical "reference". This should describe the entire functionality provided by the tool at a medium level (high level was provided by the tutorial and low level should be provided by the API docs). Currently information that might be classified as "reference" is scattered all over the book, and could definitely use some better organization.
5. Following the reference should be a conclusion linking the user to the API docs for more information.
6. Following that should be a series of appendices, if applicable. I generally see two categories of appendices: those that provide a series of technical listings that don't belong in the API docs (for example, the Rust book's [first few appendices](https://doc.rust-lang.org/book/2018-edition/appendix-01-keywords.html)), and those that provide reference for deeper parts of the tool that are not likely to be commonly used (for example, Rust's appendix on [macros](https://doc.rust-lang.org/book/2018-edition/appendix-04-macros.html)).

Given the above, I'd suggest the following structure for the docs:

* Introduction
* Getting Started
* Pong Tutorial
* Architecture
  * High Level Introduction (touch on ECS, States, what goes into an Application)
  * Potentially others
* Concepts
  * State
  * Component
  * Entity
  * Resource
  * World
  * System
  * Event Channel
  * Potentially others...

(What follows is effectively a chapter for each sub-crate of the project, in order of importance, which I might have wrong based on my limited knowledge. Some of these may belong in appendices instead of the reference chapters)
* Amethyst Core
* Renderer
* Assets
* Input
* UI
* Audio
* Animation
* Controls
* Network
* Locale
* GLTF
* Config
* Derive
* Utils
* Appendix
  * A - Amethyst Systems (provides a short description of each implementor of System and a link to it in the API docs, in order of "expected common usage")
  * B - Amethyst Bundles (same as the above but for SystemBundle implementors, including a list of Systems installed by each Bundle)
  * C - Amethyst Components (same as the above but for Component implementors)
  * D - Amethyst Resources (nearly anything can be used as a resource, but Amethyst likely has a few items that are *intended* to be used as resources, or even installs as resources itself)

The appendices here are what serve as our "catalog". It may include some duplicate information and become out of date from time to time, but I think that is to be expected, as documentation like this really can't be auto-generated. And as long as this catalog is there, people will at least be aware of a standard place to look for these items and add new items as they are discovered to be missing. Additionally, a PR template can include a checklist item akin to "Did you add a System, Bundle, Component, or Resource? Make sure to add it to the corresponding appendix in the book!"

I'm aware that the above is a lot of work, but I think the book is missing a lot of crucial information, and I'd generally prefer to have an empty skeleton that gets filled in as time goes on than an "idea" of what the book might look like some day floating around in people's minds.

**Comment by torkleyy** [comment]
You bring up many good points @jchitel, thank you for writing this. I like your proposal!

I think your reorganization is best applied in multiple steps, or how would you like to continue?
What information do you need for this exactly?

**Comment by jchitel** [comment]
I'm in the process of setting up a PR with at least some of the stuff I suggested above. But I'm obviously new to Amethyst so it's going to be a bit slow going, at least to start out.

Luckily, I'm finding the source extremely easy to read to get up to speed! The high level architecture of the engine is making a lot of sense.

**Comment by fhaynes** [comment]
Closed as this was moved to the forums for further discussion.
